spec/index.bs

```diff
@@ -8086,21 +8093,66 @@ interface GPUPresentationContext {
+        1. Let |imageContents| be
+            [$get a copy of the image contents of a presentation context|a copy of the image contents$]
+            of |context|.
+        1. Update |context|.{{GPUPresentationContext/[[canvas]]}} with |imageContents|.
+        1. Call |context|.{{GPUPresentationContext/[[currentTexture]]}}.{{GPUTexture/destroy()}}.
+    1. Set |context|.{{GPUPresentationContext/[[currentTexture]]}} to `null`.
+</div>
+
+<div algorithm="transferToImageBitmap from WebGPU">
+    When {{OffscreenCanvas/transferToImageBitmap()}} is called on a canvas with
+    {{GPUPresentationContext}} |context|:
+
+    1. Let |imageContents| be
+        [$get a copy of the image contents of a presentation context|a copy of the image contents$]
+        of |context|.
+    1. If |context|.{{GPUPresentationContext/[[currentTexture]]}} is not `null`:
+        1. Call |context|.{{GPUPresentationContext/[[currentTexture]]}}.{{GPUTexture/destroy()}}.
+    1. Set |context|.{{GPUPresentationContext/[[currentTexture]]}} to `null`.
+    1. Return a new {{ImageBitmap}} containing the |imageContents|.
 </div>
 
 <div algorithm>
-    To <dfn abstract-op lt='allocating a new presentation context texture'>Allocate a new
+    When WebGPU canvas contents are read using other Web APIs, like
+    {{CanvasDrawImage/drawImage()}}, `texImage2D()`, `texSubImage2D()`,
+    {{HTMLCanvasElement/toDataURL()}}, {{HTMLCanvasElement/toBlob()}}, and so on,
+    they <dfn abstract-op>get a copy of the image contents of a presentation context</dfn>:
+
+    **Arguments:**
+    |context|: the {{GPUPresentationContext}}
+
+    **Returns:** image contents
+
+    1. Let |texture| be |context|.{{GPUPresentationContext/[[currentTexture]]}}.
+    1. If any of the following requirements is unmet, return a transparent black image of size
+        |context|.{{GPUPresentationContext/[[size]]}} and stop.
+        <div class=validusage>
+            - |texture| must not be `null`.
+            - |texture|.{{GPUTexture/[[destroyed]]}} must be false.
+            - If |context|.{{GPUPresentationContext/[[canvas]]}} is an {{OffscreenCanvas}},
+                it must not be linked to a [=placeholder canvas element=].
+
+                Issue: If added, canvas must also not be `desynchronized`.
+        </div>
+    1. Ensure that all submitted work items (e.g. queue submissions) have
+        completed writing to |texture|.
+    1. Return the contents of |texture|.
+
+        Issue: These contents need to be tagged with the currently-configured
+        colorSpace and premultipliedAlpha mode.
+</div>
+
+Note:
+When the *user* initiates a read of a WebGPU canvas, e.g. using "Copy to Clipboard",
+"Save Image As...", printing, etc., the user agent **should** return the contents displayed on the
+screen at that moment, if possible (not the contents of {{GPUPresentationContext/[[currentTexture]]}}.
```

Comment:
Want to make sure this matches the behavior of other canvas types.  I don't think webgl or 2d canvas have a way to export the previously displayed frame since that's been handed over to the display compositor.  One way of meeting this requirement is triggering a compositing update when any of these actions are initiated so that the contents on screen matches the content copied/saved.

Reply:
I did some quick investigation and it looks like all of these actions actually cannot happen during a JS task anyway - I think they occur as events _on_ the event loop which means they can just trigger "update the rendering [of the] Document" to get the contents. At least in the case of printing, I think they must trigger it, because printing requires page relayout; I think copying probably triggers it as well.

Anyway, based on that, I don't think there's any danger of implementations returning [[currentTexture]]'s contents in the middle of a frame, so I'm just going to remove this note.